arwen/contexts/metering.go

```diff
@@ -78,38 +76,52 @@ func (context *meteringContext) BoundGasLimit(value int64) uint64 {
 	return limit
 }
 
-// DeductAndLockGasIfAsyncStep will deduct the gas for an async step and also lock gas for the callback, if the execution is an asynchronous call
-func (context *meteringContext) DeductAndLockGasIfAsyncStep() error {
-	context.gasLockedForAsyncStep = 0
-
-	input := context.host.Runtime().GetVMInput()
-	if input.CallType != vmcommon.AsynchronousCall {
-		return nil
-	}
-
+// UseGasForAsyncStep consumes the AsyncCallStep gas cost on the currently
+// running Wasmer instance
+// TODO the warmInstance will be affected if this function is misused
+func (context *meteringContext) UseGasForAsyncStep() error {
 	gasSchedule := context.GasSchedule().ElrondAPICost
+	gasToDeduct := gasSchedule.AsyncCallStep
+	return context.UseGasBounded(gasToDeduct)
+}
 
-	gasToLock := gasSchedule.AsyncCallStep + gasSchedule.AsyncCallbackGasLock
-	gasToDeduct := gasSchedule.AsyncCallStep + gasToLock
-	if input.GasProvided <= gasToDeduct {
+func (context *meteringContext) UseGasBounded(gasToUse uint64) error {
```

Comment:
wasn't use gas bounded a function which uses the maximum between gasLeft and gasToUse ? there is a problem when calling for example from a smart contract executeOnDestinationContext with all the gas. Wasmer will consume a few more gas before reaching the actual API.

code snippet from contract:
   gasLeft := eei.getGasLeft()
   executeOnDestContext("execute", gasLeft)

Reply:
That function is `metering.BoundGasLimit()`, which doesn't actually use the gas - only computes the maximum as you said. 

This function, `metering.UseGasBounded()`, has a different purpose: it actually uses the gas on the currently running Wasmer instance, and will return an early error if the gas limit is reached. This function was needed because `metering.UseGas()` did not have any check against `GasLeft()`. Maybe `UseGas()` should be obsoleted by `UseGasBounded()`.

But the names are very confusing and the difference between them is not at all obvious. Will think of better naming, thank you for pointing it out.

Reply:
UseGas in case of systemVM returns error if not enough is left. The same happens in wasmer - or when calling any API. thus we might chnge UseGas to return error if not enough gas.

Reply:
Will fix this at a later refactor.